klippa/src/cmap.rs

```diff
@@ -0,0 +1,1105 @@
+            Self::Format14(item) => item.serialize(s, plan, cp_to_new_gid_list),
+            _ => Ok(()),
+        }
+    }
+}
+
+impl Serialize for Cmap4<'_> {
+    fn serialize(
+        &self,
+        s: &mut Serializer,
+        _plan: &Plan,
+        cp_to_new_gid_list: &[(u32, GlyphId)],
+    ) -> Result<(), SerializeErrorFlags> {
+        if cp_to_new_gid_list.is_empty() {
+            return Ok(());
+        }
+
+        let init_len = s.length();
+        // format
+        s.embed(4_u16)?;
+        // length, initialized to 0
+        let length_pos = s.embed(0_u16)?;
+        // language
+        s.embed(self.language())?;
+        // segCountx2, initialized to 0
+        let segcount_pos = s.embed(0_u16)?;
+        // searchRange, initialized to 0
+        let search_range_pos = s.embed(0_u16)?;
+        // entrySelector, initialized to 0
+        let entry_selector_pos = s.embed(0_u16)?;
+        // rangeShift, initialized to 0
+        let rangeshift_pos = s.embed(0_u16)?;
+
+        let seg_count = serialize_find_segcount(s, cp_to_new_gid_list);
+        let (end_code, start_code, id_delta) =
+            serialize_start_end_delta_arrays(s, cp_to_new_gid_list, seg_count as usize)?;
+        serialize_rangeoffset_glyph_ids(
+            s,
+            cp_to_new_gid_list,
+            seg_count as usize,
+            end_code,
+            start_code,
+            id_delta,
+        )?;
+
+        s.check_assign::<u16>(
+            length_pos,
+            s.length() - init_len,
+            SerializeErrorFlags::SERIALIZE_ERROR_INT_OVERFLOW,
+        );
+        s.check_assign::<u16>(
+            segcount_pos,
+            seg_count as usize * 2,
+            SerializeErrorFlags::SERIALIZE_ERROR_INT_OVERFLOW,
+        );
+
+        let entry_selector = 1_u16.max(16 - seg_count.leading_zeros() as u16) - 1;
+        s.copy_assign(entry_selector_pos, entry_selector);
+
+        let search_range = 2 * (1 << entry_selector) as usize;
+        s.check_assign::<u16>(
+            search_range_pos,
+            search_range,
+            SerializeErrorFlags::SERIALIZE_ERROR_INT_OVERFLOW,
+        );
+
+        let range_shift = if seg_count as usize * 2 > search_range {
+            seg_count * 2 - search_range as u16
+        } else {
+            0
+        };
+        s.copy_assign(rangeshift_pos, range_shift);
+
+        Ok(())
+    }
+}
+
+trait UnicodeRangeWriter {
+    /// write or count ranges
+    fn write_one_range(&mut self, start: u16, end: u16, delta: i16, s: &mut Serializer);
+}
+
+#[derive(Default)]
+struct RangeCounter {
+    seg_count: u16,
+}
+
+impl RangeCounter {
+    fn new() -> Self {
+        RangeCounter::default()
+    }
+
+    fn count(&self) -> u16 {
+        self.seg_count
+    }
+}
+
+impl UnicodeRangeWriter for RangeCounter {
+    fn write_one_range(&mut self, _start: u16, _end: u16, _delta: i16, _s: &mut Serializer) {
+        self.seg_count += 1;
+    }
+}
+
+fn to_ranges(
```

Comment:
if this is ported from something in harfbuzz can we include a link? (and if not, it deserves some documentation, it's a non-trivial fn)